### PR TITLE
[Issue-278] Add check for BsonNull in processDocumentsInArray and processArrayInArray

### LINKED ISSUE #278: Value expected to be of type DOCUMENT is of unexpected type NULL
## Environment
- DocumentDB JDBC driver version: 1.14.0.0
- DocumentDB server version: 3.6
- OS: [e.g. MacOS Big Sur 11.5.1, Windows 10 64-bit]: Windows 10 64-bit
- BI Tool or client name:  Tableau
- BI Tool or client version: 2021.4.3
- Java version (if known): unkown

---

## Problem Description
1. Steps to reproduce: 
* Copy the jdbc driver "1.1.1-all.jar" to `C:\Program Files\Tableau\Drivers`
* Copy the .taco "documentdbjdbc-1.1.1.taco" to `C:\Users\[user]\Documents\My Tableau Repository\Connectors`
* Open tableau and try to create a new DocumentDb Connection. 
* When clicking on "sign in" the following error occurs:
```
An error occurred while communicating with Amazon DocumentDB by Amazon Web Services
Bad Connection: Tableau could not connect to the data source.
Error Code: FAB9A2C5
Value expected to be of type DOCUMENT is of unexpected type NULL
Unable to connect to the Amazon DocumentDB by Amazon Web Services server "indivupad.cluster-REDACTED.us-east-1.docdb.amazonaws.com". Check that the server is running and that you have access privileges to the requested database.
Connector Class: documentdbjdbc, Version: 1.14.0.0
```
2. Expected behaviour: connection would establish successful
3. Actual behaviour: connection didnt establish
4. Error message/stack trace:
see above

---

## Screenshots

---

## JDBC log
```
2022-02-14 12:12:53.680 +0100 (,,,,10,17) grpc-default-executor-7 : ERROR com.tableau.connect.util.GrpcServiceHelper - Failed in constructProtocol.
shadow.org.bson.BsonInvalidOperationException: Value expected to be of type DOCUMENT is of unexpected type NULL
	at shadow.org.bson.BsonValue.throwIfInvalidType(BsonValue.java:419) ~[?:?]
	at shadow.org.bson.BsonValue.asDocument(BsonValue.java:47) ~[?:?]
	at software.amazon.documentdb.jdbc.metadata.DocumentDbTableSchemaGenerator.processDocumentsInArray(DocumentDbTableSchemaGenerator.java:540) ~[?:?]
	at software.amazon.documentdb.jdbc.metadata.DocumentDbTableSchemaGenerator.processArray(DocumentDbTableSchemaGenerator.java:405) ~[?:?]
	at software.amazon.documentdb.jdbc.metadata.DocumentDbTableSchemaGenerator.processComplexTypes(DocumentDbTableSchemaGenerator.java:562) ~[?:?]
	at software.amazon.documentdb.jdbc.metadata.DocumentDbTableSchemaGenerator.processDocument(DocumentDbTableSchemaGenerator.java:188) ~[?:?]
	at software.amazon.documentdb.jdbc.metadata.DocumentDbTableSchemaGenerator.processDocumentsInArray(DocumentDbTableSchemaGenerator.java:540) ~[?:?]
	at software.amazon.documentdb.jdbc.metadata.DocumentDbTableSchemaGenerator.processArray(DocumentDbTableSchemaGenerator.java:405) ~[?:?]
	at software.amazon.documentdb.jdbc.metadata.DocumentDbTableSchemaGenerator.processComplexTypes(DocumentDbTableSchemaGenerator.java:562) ~[?:?]
	at software.amazon.documentdb.jdbc.metadata.DocumentDbTableSchemaGenerator.processDocument(DocumentDbTableSchemaGenerator.java:188) ~[?:?]
	at software.amazon.documentdb.jdbc.metadata.DocumentDbTableSchemaGenerator.processComplexTypes(DocumentDbTableSchemaGenerator.java:557) ~[?:?]
	at software.amazon.documentdb.jdbc.metadata.DocumentDbTableSchemaGenerator.processDocument(DocumentDbTableSchemaGenerator.java:188) ~[?:?]
	at software.amazon.documentdb.jdbc.metadata.DocumentDbTableSchemaGenerator.processComplexTypes(DocumentDbTableSchemaGenerator.java:557) ~[?:?]
	at software.amazon.documentdb.jdbc.metadata.DocumentDbTableSchemaGenerator.processDocument(DocumentDbTableSchemaGenerator.java:188) ~[?:?]
	at software.amazon.documentdb.jdbc.metadata.DocumentDbTableSchemaGenerator.generate(DocumentDbTableSchemaGenerator.java:80) ~[?:?]
	at software.amazon.documentdb.jdbc.metadata.DocumentDbMetadataService.getCollectionMetadataDirect(DocumentDbMetadataService.java:404) ~[?:?]
	at software.amazon.documentdb.jdbc.metadata.DocumentDbMetadataService.getNewDatabaseMetadata(DocumentDbMetadataService.java:354) ~[?:?]
	at software.amazon.documentdb.jdbc.metadata.DocumentDbMetadataService.get(DocumentDbMetadataService.java:128) ~[?:?]
	at software.amazon.documentdb.jdbc.metadata.DocumentDbDatabaseSchemaMetadata.get(DocumentDbDatabaseSchemaMetadata.java:114) ~[?:?]
	at software.amazon.documentdb.jdbc.DocumentDbConnection.setMetadata(DocumentDbConnection.java:234) ~[?:?]
	at software.amazon.documentdb.jdbc.DocumentDbConnection.ensureDatabaseMetadata(DocumentDbConnection.java:229) ~[?:?]
	at software.amazon.documentdb.jdbc.DocumentDbConnection.getMetaData(DocumentDbConnection.java:208) ~[?:?]
	at com.tableausoftware.jdbc.JDBCProtocolImpl.getDriverName(JDBCProtocolImpl.java:1037) ~[jdbcserver.jar:20214.0.17]
	at com.tableau.connect.service.ProtocolService.constructProtocol(ProtocolService.java:61) ~[jdbcserver.jar:20214.0.17]
	at com.tableau.connect.grpc.GrpcProtocolService.lambda$constructProtocol$0(GrpcProtocolService.java:63) ~[jdbcserver.jar:20214.0.17]
	at com.tableau.connect.grpc.GrpcProtocolService.wrap(GrpcProtocolService.java:289) ~[jdbcserver.jar:20214.0.17]
	at com.tableau.connect.grpc.GrpcProtocolService.constructProtocol(GrpcProtocolService.java:62) ~[jdbcserver.jar:20214.0.17]
	at com.tableau.connect.generated.ProtocolServiceGrpc$MethodHandlers.invoke(ProtocolServiceGrpc.java:1492) ~[jdbcserver.jar:20214.0.17]
	at io.grpc.stub.ServerCalls$UnaryServerCallHandler$UnaryServerCallListener.onHalfClose(ServerCalls.java:180) ~[jdbcserver.jar:20214.0.17]
	at io.grpc.PartialForwardingServerCallListener.onHalfClose(PartialForwardingServerCallListener.java:35) ~[jdbcserver.jar:20214.0.17]
	at io.grpc.ForwardingServerCallListener.onHalfClose(ForwardingServerCallListener.java:23) ~[jdbcserver.jar:20214.0.17]
	at io.grpc.ForwardingServerCallListener$SimpleForwardingServerCallListener.onHalfClose(ForwardingServerCallListener.java:40) ~[jdbcserver.jar:20214.0.17]
	at io.grpc.Contexts$ContextualizedServerCallListener.onHalfClose(Contexts.java:86) ~[jdbcserver.jar:20214.0.17]
	at io.grpc.internal.ServerCallImpl$ServerStreamListenerImpl.halfClosed(ServerCallImpl.java:331) ~[jdbcserver.jar:20214.0.17]
	at io.grpc.internal.ServerImpl$JumpToApplicationThreadServerStreamListener$1HalfClosed.runInContext(ServerImpl.java:814) ~[jdbcserver.jar:20214.0.17]
	at io.grpc.internal.ContextRunnable.run(ContextRunnable.java:37) ~[jdbcserver.jar:20214.0.17]
	at io.grpc.internal.SerializingExecutor.run(SerializingExecutor.java:123) ~[jdbcserver.jar:20214.0.17]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source) ~[?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source) ~[?:?]
```

### PR DESCRIPTION
**Pull request by andiemontoyeah**
### Summary

- Resolves #278 

### Description

- Added check for null in `processDocumentsInArray` and `processArrayInArray`
- Added unit tests
- Changed if statement in `processArray` to use table name not path since it is the name being used as key - this broke a test until moving some of the logic outside the if statement.

### Related Issue

#278 
https://bitquill.atlassian.net/browse/AD-602
### Additional Reviewers
@affonsoBQ
@andiem-bq
@alexey-temnikov
@birschick-bq

### CODE REVIEW

File: calcite-adapter/src/main/java/software/amazon/documentdb/jdbc/metadata/DocumentDbTableSchemaGenerator.java [hunk 1]
```diff
@@ -324,7 +324,7 @@ private static void processArray(
             handleComplexScalarConflict(tableMap, tableName, columnMap);
         }
 
-        if (!tableMap.containsKey(path)) {
+        if (!tableMap.containsKey(tableName)) {
```
Comment (andiemontoyeah) [inline calcite-adapter/src/main/java/software/amazon/documentdb/jdbc/metadata/DocumentDbTableSchemaGenerator.java:327]: changed this check to match the rest - the code in here was always being executed since tableMap uses the table names as the key

File: calcite-adapter/src/main/java/software/amazon/documentdb/jdbc/metadata/DocumentDbTableSchemaGenerator.java [hunk 2]
```diff
@@ -363,41 +363,41 @@ private static void processArray(
...
-                        .isGenerated(true)
-                        .build();
-                columnMap.put(indexColumn.getSqlName(), indexColumn);
-            } else {
-                // Cast exception should not occur, because we are always creating DocumentDbMetadataColumn.
-                indexColumn = (DocumentDbMetadataColumn) columnMap.get(indexColumnName);
-            }
-            // Add index column to foreign keys
-            foreignKeys.add(indexColumn);
+        final Map<String, String> columnNameMap = columnMap.values().stream().collect(
```
Comment (andiemontoyeah) [inline calcite-adapter/src/main/java/software/amazon/documentdb/jdbc/metadata/DocumentDbTableSchemaGenerator.java:368]: after changing from using path as  the key to using the table name, index was not being added to foreign key list for nested array in array or nested document in array for the case where the nested document/array in array is not in first document scanned. Moved adding index to foreign key to outside the if statement.

File: calcite-adapter/src/test/java/software/amazon/documentdb/jdbc/metadata/DocumentDbTableSchemaGeneratorTest.java [hunk 3]
```diff
@@ -617,7 +617,7 @@ void testComplexSingleLevelArray() {
     void testComplexTwoLevelArray() {
         final Map<String, String> tableNameMap = new HashMap<>();
         final BsonDocument document = BsonDocument.parse(
-                "{ \"_id\" : \"key\", \"array\" : [ [1, 2, 3 ], [ 4, 5, 6 ] ]}");
+                "{ \"_id\" : \"key\", \"array\" : [ [1, 2, 3 ], [ 4, 5, 6 ], [7, 8, 9 ] ]}");
```
Comment (andiemontoyeah) [inline calcite-adapter/src/test/java/software/amazon/documentdb/jdbc/metadata/DocumentDbTableSchemaGeneratorTest.java]: Added additional nested array entry here. Before changing if statement to use table name and moving adding foreign key to outside if statement, the second array index was being added to foreign key list multiple times. this wasnt adding extra columns since a map is used BUT primary key index could be over incremented. With testing with this extra entry initially, the primary key index was showing 4 instead of 3.

File: calcite-adapter/src/test/java/software/amazon/documentdb/jdbc/metadata/DocumentDbTableSchemaGeneratorTest.java [hunk 4]
```diff
@@ -1077,15 +1077,15 @@ void testInconsistentArrayDataType() {
         final Map<String, String> tableNameMap = new HashMap<>();
         final BsonDocument[] tests = new BsonDocument[] {
                 BsonDocument.parse(
-                        "{ \"_id\" : \"key\", \"array\" : [ 1, [ 2 ] ] }"),
+                        "{ \"_id\" : \"key\", \"array\" : [ 1, [ 2 ], null ] }"),
```
Comment (andiemontoyeah) [inline calcite-adapter/src/test/java/software/amazon/documentdb/jdbc/metadata/DocumentDbTableSchemaGeneratorTest.java]: Added instances of null to this test just to make sure that case is covered

File: calcite-adapter/src/test/java/software/amazon/documentdb/jdbc/metadata/DocumentDbTableSchemaGeneratorTest.java [hunk 5]
```diff
@@ -1753,6 +1753,220 @@ void testSubDocumentInArrayOnlyInSecondDocument() {
         Assertions.assertNotNull(table2.getColumnMap().get("data_locations_index_lvl_0"));
     }
 
+    @DisplayName("Tests an array with multiple documents and nulls.")
+    @Test
+    void testArrayWithDocumentsAndNull() {
```
Comment (andiemontoyeah) [inline calcite-adapter/src/test/java/software/amazon/documentdb/jdbc/metadata/DocumentDbTableSchemaGeneratorTest.java]: This case used to cause an exception in processDocumentsInArray

File: calcite-adapter/src/test/java/software/amazon/documentdb/jdbc/metadata/DocumentDbTableSchemaGeneratorTest.java [hunk 6]
```diff
@@ -1753,6 +1753,220 @@ void testSubDocumentInArrayOnlyInSecondDocument() {
...
+        Assertions.assertFalse(metadataColumn.isGenerated());
+
+        printMetadataOutput(metadata, getMethodName());
+    }
+
+    @DisplayName("Tests a two-level scalar array with nulls on both levels.")
+    @Test
+    void testTwoLevelArrayWithNulls() {
+        final Map<String, String> tableNameMap = new HashMap<>();
+        final BsonDocument document = BsonDocument.parse(
```
Comment (andiemontoyeah) [inline calcite-adapter/src/test/java/software/amazon/documentdb/jdbc/metadata/DocumentDbTableSchemaGeneratorTest.java]: This case caused similar exception in processArrayInArray

File: src/test/java/software/amazon/documentdb/jdbc/DocumentDbStatementJoinTest.java [hunk 7]
```diff
@@ -38,11 +38,13 @@
     private static final String MULTI_NESTED_ARRAY_AND_DOCUMENT_JSON = "{ \"_id\" : \"key1\",  "
             + "\"document\": {\"field\": 1},"
             + "\"array\" : ["
-            + " {\"field\": 1, \"document\": {\"field\": 1} , \"array\": [1, 2, 3], \"otherArray\": [4, 5, 6]}, "
-            + " {\"field\": 2, \"document\": {\"field\": 2} , \"array\": [4, 5, 6], \"otherArray\": [7, 8, 9]} ], "
+            + " {\"field\": 1, \"document\": {\"field\": 1} , \"array\": [1, 2, 3, null], \"otherArray\": [4, 5, 6]}, "
```
Comment (andiemontoyeah) [inline src/test/java/software/amazon/documentdb/jdbc/DocumentDbStatementJoinTest.java:41]: Adding instances of null to test data for same collection join tests

File: calcite-adapter/src/test/java/software/amazon/documentdb/jdbc/metadata/DocumentDbTableSchemaGeneratorTest.java [hunk 8]
```diff
@@ -1077,15 +1077,15 @@ void testInconsistentArrayDataType() {
...
+                        "{ \"_id\" : \"key\", \"array\" : [ 1, [ 2 ], null ] }"),
                 BsonDocument.parse(
                         "{ \"_id\" : \"key\", \"array\" : [ 1, { \"field\" : 2 } ] }"),
                 BsonDocument.parse(
                         "{ \"_id\" : \"key\", \"array\" : [ { \"field\" : 2 }, 3 ] }"),
                 BsonDocument.parse(
                         "{ \"_id\" : \"key\", \"array\" : [ { \"field\" : 2 }, [ 3, 4 ] ] }"),
                 BsonDocument.parse(
-                        "{ \"_id\" : \"key\", \"array\" : [ [ 1, 2 ], { \"field\" : 2 } ] }"),
+                        "{ \"_id\" : \"key\", \"array\" : [ [ 1, 2 ], { \"field\" : 2 }, null ] }"),
```
Comment (alexey-temnikov) [inline calcite-adapter/src/test/java/software/amazon/documentdb/jdbc/metadata/DocumentDbTableSchemaGeneratorTest.java]: Do we have a test where the field value might be null or another array with a null element?

Comment (andiemontoyeah) [reply calcite-adapter/src/test/java/software/amazon/documentdb/jdbc/metadata/DocumentDbTableSchemaGeneratorTest.java]: Not for nested arrays. Will add! 👍

**Review by andiemontoyeah — COMMENTED** [review]
Got into the pickle of DocumentDbTableSchemaGeneratorTest being too many lines - separated these into 2 different files. :( 
BasicTest has the relatively simpler test cases 
ComplexTest has the ones added later to address edge cases 
Let me know if a different split makes more sense!

File: calcite-adapter/src/test/java/software/amazon/documentdb/jdbc/metadata/DocumentDbTableSchemaGeneratorBasicTest.java [hunk 9]
```diff
@@ -0,0 +1,1471 @@
...
+
+        printMetadataOutput(metadata, getMethodName());
+    }
+
+    /**
+     * Tests a two-level document.
+     */
+    @DisplayName("Tests a two-level document.")
+    @Test
+    void testComplexTwoLevelDocument() {
```
Comment (alinaliBQ) [inline calcite-adapter/src/test/java/software/amazon/documentdb/jdbc/metadata/DocumentDbTableSchemaGeneratorBasicTest.java]: It can be confusing when there are some test with the word "complex" in their names inside "DocumentDbTableSchemaGenerator**Basic**Test". Maybe name "DocumentDbTableSchemaGenerator**Complex**Test" to "DocumentDbTableSchemaGenerator**Adanced**Test" ?

Comment (affonsov) [reply calcite-adapter/src/test/java/software/amazon/documentdb/jdbc/metadata/DocumentDbTableSchemaGeneratorBasicTest.java]: Using Basic/Complex/Advanced can be very subjective.
It is better to use other criteria to split the class, or add a comment into the class level explaining what is considered basic and what is complex

Comment (andiemontoyeah) [reply calcite-adapter/src/test/java/software/amazon/documentdb/jdbc/metadata/DocumentDbTableSchemaGeneratorBasicTest.java]: Agree this is pretty random and confusing. 😅  Tried to split in a new way, ending up with 3 new files: 
- Identifier - checks identifier truncation logic for column and table names
- VirtualTable - more focused on testing if we get correct set of virtual tables and foreign keys from documents with nested structure
- Column - more focused on testing correct column type promotion

**Review by alinaliBQ — APPROVED** [review]
lgtm!

**Review by birschick-bq — APPROVED** [review]

**Review by affonsov — APPROVED** [review]
LGTM
